File: video_router/pool.py
"""
pool.py — M1 iş havuzu (AYRI KATMAN: runner + L2.5 kullanır).
================================================================================
Tasarım (A): adaptör `submit()` BÜTÜN olarak worker thread'de koşar (upload→generate→poll→download).
Poll-bekleme örtüşür → üretim PARALEL (B testi: ~1.6-2×). concurrency=1 → sıralıya denk (regresyon).

- concurrency  : eşzamanlılık tavanı. Firefly=1, Hailuo=2-4 (ölçülen). ThreadPool(N) = BİRİNCİL fren.
                 (wait_for_queue Pool'da ATLANIR — kendi işlerimizi sayıp self-deadlock yapardı.)
- store        : ProgressStore (tek-yazıcı) — on_submit/done/failed KİLİTLİ + ATOMİK.
- _gate        : GLOBAL submit kapısı (pacing) — son generate'ten ≥rastgele(20-60)s geçmeden yenisi gitmez;
                 worker'lar tek tek geçer (yığılma/bot koruması). Adaptör generate'ten önce job.pre_generate() çağırır.
- on_submit    : iş-başına — submit OK olunca vid_id ANINDA store'a (kesintide resume kurtarır).
- hata izolasyonu: patlayan iş JobResult(ok=False) + record_failure (vid_id korunur); havuz DEVAM.
- sırasız biter; çağıran (runner) index'e göre sıralar.
"""

import time
import random
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from typing import Optional, Callable


@dataclass
class JobResult:
    job: object
    ok: bool
    path: object = None
    error: Optional[str] = None
    used_spec: object = None
    meta: Optional[dict] = None


class Pool:
    def __init__(self, concurrency: int, store, pacing=(20, 60)):
        self.concurrency = max(1, int(concurrency))
        self.store = store
        self.pacing = pacing
        self._gate_lock = threading.Lock()
        self._last_generate = 0.0

    def _gate(self):
        """GLOBAL submit kapısı: son generate'ten ≥rastgele(pacing)s geçene dek bekle. Kilitli = tek tek geçiş."""
        lo, hi = self.pacing
        with self._gate_lock:
            if hi and hi > 0:
                wait = random.randint(lo, hi)
                if self._last_generate:
                    elapsed = time.time() - self._last_generate
                    if elapsed < wait:
                        time.sleep(wait - elapsed)
            self._last_generate = time.time()

    def _work(self, job, produce):
        # Pool kancaları (adaptör bunları okur)
        job.skip_queue_gate = True
        job.pre_generate = self._gate
        job.on_submit = lambda vid, j=job: self.store.set_submitted(j.out_name, vid, j.submit_meta or {})
        base = dict(job.submit_meta or {})
        try:
            path, used_spec, meta = produce(job)
            rec = dict(base); rec["status"] = "done"
            if meta:                       # file/softened/fallback... duz-merge -> sirali done kaydiyla ayni
                rec.update(meta)
            self.store.update(job.out_name, rec)
            return JobResult(job, True, path, None, used_spec, meta)
        except (Exception, SystemExit) as e:
            rec = dict(base); rec["status"] = "error"; rec["error"] = str(e)
            self.store.record_failure(job.out_name, rec)   # vid_id KORUNUR
            return JobResult(job, False, None, str(e))

    def run(self, jobs, produce, on_result=None):
        """jobs listesini işle. produce(job)->(path,spec,meta) [S4 dahil]. Sonuç listesi (sırasız biter)."""
        results = []
        if self.concurrency == 1:
            # SIRALI (regresyon): ThreadPool overhead'siz, aynı work mantığı.
            for job in jobs:
                r = self._work(job, produce)
                if on_result:
                    on_result(r)
                results.append(r)
            return results
        with ThreadPoolExecutor(max_workers=self.concurrency) as ex:
            futs = {ex.submit(self._work, j, produce): j for j in jobs}
            for f in as_completed(futs):
                r = f.result()
                if on_result:
                    on_result(r)        # ANA thread (çağıranın kilit derdi yok)
                results.append(r)
        return results

File: src/lib/pipeline/router/pool.ts
import type { AdapterSpec, Job } from "./core";
import type { ProgressStore } from "./progress";

export type JobResult = {
  job: PoolJob;
  ok: boolean;
  path?: string;
  error?: string;
  usedSpec?: AdapterSpec;
  meta?: Record<string, unknown>;
};

export type PoolJob = Job & {
  _spec?: AdapterSpec;
  _mode?: string;
  outName?: string;
  submitMeta?: Record<string, unknown>;
};

export type ProduceFn = (job: PoolJob) => Promise<{
  path: string;
  usedSpec: AdapterSpec;
  meta: Record<string, unknown>;
}>;

export class Pool {
  private lastGenerate = 0;
  private gateLock: Promise<void> = Promise.resolve();

  constructor(
    public concurrency: number,
    private store: ProgressStore,
    private pacing: [number, number] = [20, 60],
  ) {
    this.concurrency = Math.max(1, concurrency);
  }

  private async gate() {
    const [lo, hi] = this.pacing;
    let release!: () => void;
    const waitTurn = new Promise<void>((r) => {
      release = r;
    });
    const prev = this.gateLock;
    this.gateLock = waitTurn;
    await prev;

    try {
      if (hi > 0) {
        const wait = lo + Math.floor(Math.random() * (hi - lo + 1));
        if (this.lastGenerate) {
          const elapsed = (Date.now() - this.lastGenerate) / 1000;
          if (elapsed < wait) {
            await new Promise((r) => setTimeout(r, (wait - elapsed) * 1000));
          }
        }
      }
      this.lastGenerate = Date.now();
    } finally {
      release();
    }
  }

  private async work(job: PoolJob, produce: ProduceFn): Promise<JobResult> {
    job.skipQueueGate = true;
    job.preGenerate = () => this.gate();
    job.onSubmit = async (vid) => {
      await this.store.setSubmitted(job.outName!, vid, job.submitMeta || {});
    };
    const base = { ...(job.submitMeta || {}) };
    try {
      const { path, usedSpec, meta } = await produce(job);
      const rec = { ...base, status: "done", ...meta };
      await this.store.update(job.outName!, rec);
      return { job, ok: true, path, usedSpec, meta };
    } catch (e) {
      const rec = { ...base, status: "error", error: String(e) };
      await this.store.recordFailure(job.outName!, rec);
      return { job, ok: false, error: String(e) };
    }
  }

  async run(
    jobs: PoolJob[],
    produce: ProduceFn,
    onResult?: (r: JobResult) => void,
  ): Promise<JobResult[]> {
    const results: JobResult[] = [];
    if (this.concurrency === 1) {
      for (const job of jobs) {
        const r = await this.work(job, produce);
        onResult?.(r);
        results.push(r);
      }
      return results;
    }

    let idx = 0;
    const worker = async () => {
      while (true) {
        const i = idx++;
        if (i >= jobs.length) break;
        const r = await this.work(jobs[i], produce);
        onResult?.(r);
        results.push(r);
      }
    };
    await Promise.all(Array.from({ length: this.concurrency }, worker));
    return results;
  }
}
